### PR TITLE
added id and type filter for hooks

### LINKED ISSUE #341: Group hooks by type
It would be really handy to have the hooks grouped by what it specified in the `types` field in the JSON data. That way if I'm looking to see what hooks apply to Python I can look in that section. I would assume this would be its own `<h2>` section after the current by-repository list.

### PR DESCRIPTION
**Pull request by paulhfischer**
resolves #341 

- options for select are automatically generated and only include types for hooks that are in `all-hooks.json`
- also added a search input for the hook id
- repos with non-matching hooks are completely hidden

### CODE REVIEW

File: hooks.mako [hunk 1]
```diff
@@ -1,7 +1,18 @@
...
-<div class="page-header"><h1>Supported hooks</h1></div>
+<div class="page-header">
+    <h1>Supported hooks</h1>
+    <div class="filter">
+        <input id="search-hook-id" type="text" placeholder="filter hook id"></input>
+        <select id="select-hook-type">
+            <option value="">filter hook type</option>
+            % for hook_type in all_types:
+                <option value="${hook_type}">${hook_type}</option>
+            % endfor
```
Comment (asottile) [inline hooks.mako:12]: these should be sorted

File: hooks.mako [hunk 2]
```diff
@@ -1,7 +1,18 @@
 ## -*- coding: utf-8 -*-
 <%inherit file="base.mako" />
 
-<div class="page-header"><h1>Supported hooks</h1></div>
+<div class="page-header">
+    <h1>Supported hooks</h1>
+    <div class="filter">
+        <input id="search-hook-id" type="text" placeholder="filter hook id"></input>
```
Comment (asottile) [inline hooks.mako:7]: for the smallest responsive size, this needs more margin

I wonder if there's a bootstrap forms thing that does all this for you

File: assets/filter_repos.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+            }
+        }
+
+        repo.hidden = !hasVisibleHooks;
+        hooks.querySelector(`h3[data-repo="${repo.dataset.repo}"]`).hidden = !hasVisibleHooks;
+    }
+};
+
+searchInput.addEventListener('input', filterHooks);
+selectInput.addEventListener('change', filterHooks);
```
Comment (asottile) [inline assets/filter_repos.js:37]: will need to look at caniuse for:
- `let`
- fat arrow
- format strings
- `const`
- `.includes(...)`
- `.dataset`

I suspect at least one of those things will need to be made more-compatible

File: assets/filter_repos.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
+const searchInput = document.getElementById('search-hook-id');
+const selectInput = document.getElementById('select-hook-type');
+
+const hooks = document.getElementById('hooks');
+const repos = hooks.getElementsByTagName('ul');
```
Comment (asottile) [inline assets/filter_repos.js:5]: this sets all of these as globals -- this should be wrapped in an anonymous function (+ `'use strict';`)

File: hooks.mako [hunk 5]
```diff
@@ -1,7 +1,18 @@
...
+        <select id="select-hook-type">
```
Comment (asottile) [inline hooks.mako:8]: for some reason this select looks really broken on windows

would it be better to have labels for both of these inputs and then have the placeholder value here be `--` or something?  might make this a little nicer

File: make_templates.py [hunk 6]
```diff
@@ -25,7 +25,12 @@ def get_env() -> Dict[str, Any]:
         open('all-hooks.json').read(),
         object_pairs_hook=collections.OrderedDict,
     )
-    return {'all_hooks': all_hooks}
+    all_types = {
+        hook_type
+        for properties in all_hooks.values()
+        for hook_type in properties[0].get("types", [])
```
Comment (asottile) [inline make_templates.py:31]: there's also `types_or` now, not sure how this will filter 🤔

File: scss/main.scss [hunk 7]
```diff
@@ -225,3 +225,31 @@ h3 {
...
+    select {
+        padding: 6px 12px;
+        box-sizing: border-box;
+        height: 38px;
+    }
+
+    input {
+        width: 250px;
+    }
+}
```
Comment (asottile) [inline scss/main.scss:255]: will need to caniuse for:

- flexbox
- justify-content: space-between
- :not(...)
- :last-child (though iirc this has wide support)

File: scss/main.scss [hunk 8]
```diff
@@ -225,3 +225,31 @@ h3 {
         display: inline;
     }
 }
+
+.page-header {
```
Comment (asottile) [inline scss/main.scss:229]: I think most of this css can be replaced with the bootstrap classes probably

File: assets/filter_repos.js (hunk 3 above)
Comment (paulhfischer) [reply assets/filter_repos.js:37]: the whole code fits the ES6, detailed caniuse:

* `let`: [96.75%](https://caniuse.com/let)
* fat arrow: [95.47%](https://caniuse.com/arrow-functions)
* format strings: [95.99%](https://caniuse.com/template-literals)
* `const`: [99.16%](https://caniuse.com/const)
* `.includes(...)`: [95.98%](https://caniuse.com/es6-string-includes)
* `.dataset`: [99.54%](https://caniuse.com/dataset)

to meet `>0.2%, not dead`, `const` and `let` have to be changed to var and the template literal with e.g. `.concat`.

File: hooks.mako (hunk 2 above)
Comment (paulhfischer) [reply hooks.mako:7]: i restructured the header with bootstrap `col` and `row`.

File: hooks.mako (hunk 5 above)
Comment (paulhfischer) [reply hooks.mako:8]: i think this should be fixed as i now use the bootstrap classes for styling the inputs.

added labels and moved the inputs out of the header, as the header would look really cluttered imo (and too high compared to the other sites due to the labels).

File: make_templates.py (hunk 6 above)
Comment (paulhfischer) [reply make_templates.py:31]: i joined `types` and `types_or` now.

File: scss/main.scss (hunk 7 above)
Comment (paulhfischer) [reply scss/main.scss:255]: i was able to get rid of all those items as i moved the inputs out of the header.

File: scss/main.scss (hunk 8 above)
Comment (paulhfischer) [reply scss/main.scss:229]: so the problem is that the site currently depends on bootstrap 3, which doesn't support the `d-flex` etc. classes. i can update it to bootstrap 4, this however causes an error when being compiled with `pysassc` (thus, we might have to switch to another compiler).

let me know if i should do this update, for now i renamed the classes to match the naming of the needed bootstrap 4 classes.

for now i renamed the flexbox classes to match the naming of the bootstrap 4 ones.

Comment (asottile) [reply scss/main.scss:229]: it isn't a problem with the compiler, there's just a bunch of breaking changes between 3 and 4 (I tried migrating at some point).  bs4 works fine with `pysassc` (it's the same as node-sass) -- I maintain it btw

**Comment by paulhfischer** [comment]
(i will squash the commits later, but it might be easier to see what i changed with a separate commit)

File: scss/main.scss (hunk 8 above)
Comment (paulhfischer) [reply scss/main.scss]: ah, okay, didn't look at the error message in detail tbh (as it's also possible with bs3). the error was reported in line 2 which is the bootstrap import, so i guessed the error was caused by bootstrap itself (but it's not).

**Review by asottile — APPROVED** [review]

**Pull request merged**